src/ray/raylet/node_manager.cc

```diff
@@ -1765,79 +1782,131 @@ bool NodeManager::AssignTask(const Task &task) {
-            auto actor_entry = actor_registry_.find(spec.ActorId());
-            RAY_CHECK(actor_entry != actor_registry_.end());
-            // Process any new actor handles that were created since the
-            // previous task on this handle was executed. The first task
-            // submitted on a new actor handle will depend on the dummy object
-            // returned by the previous task, so the dependency will not be
-            // released until this first task is submitted.
-            for (auto &new_handle_id : spec.NewActorHandles()) {
-              // Get the execution dependency for the first task submitted on the new
-              // actor handle. Since the new actor handle was created after this task
-              // began and before this task finished, it must have the same execution
-              // dependency.
-              const auto &execution_dependencies =
-                  assigned_task.GetTaskExecutionSpec().ExecutionDependencies();
-              // TODO(swang): We expect this task to have exactly 1 execution dependency,
-              // the dummy object returned by the previous actor task. However, this
-              // leaks information about the TaskExecutionSpecification implementation.
-              RAY_CHECK(execution_dependencies.size() == 1);
-              const ObjectID &execution_dependency = execution_dependencies.front();
-              // Add the new handle and give it a reference to the finished task's
-              // execution dependency.
-              actor_entry->second.AddHandle(new_handle_id, execution_dependency);
-            }
-
-            // TODO(swang): For actors with multiple actor handles, to
-            // guarantee that tasks are replayed in the same order after a
-            // failure, we must update the task's execution dependency to be
-            // the actor's current execution dependency.
-          } else {
-            RAY_CHECK(spec.NewActorHandles().empty());
+  if (worker->Port() > 0) {
+    // Worker is listening on a port. Push the task directly to worker.
+    // Lookup worker client for this node_id and use it to send the request.
+    auto client_id = worker->Connection()->GetClientId();
+    auto client_entry = worker_clients_.find(client_id);
+    if (client_entry == worker_clients_.end()) {
+      // TODO(atumanov): caller must handle failure to ensure tasks are not lost.
+      RAY_LOG(INFO) << "No worker found for client id " << client_id;
+      return false;
+    }
+    auto &client = client_entry->second;
+
+    rpc::AssignTaskRequest request;
+    request.set_task_id(task_id.Binary());
+    request.set_task_spec(task.Serialize());
+    request.set_resource_ids(resource_id_set.Serialize());
+
+    auto status = client->AssignTask(
+        request, [this, worker](Status status, const rpc::AssignTaskReply &reply) {
+          // Worker has finished this task. The logic below is the same
+          // as ProcessGetTaskMessage().
```

Comment:
Why not call `ProcessGetTaskMessage()` directly? It would be better if we can share as much code as possible with the `GetTask` codepath.

Reply:
revised.